### ISSUE TITLE
MultiRNNCell error

### ISSUE THREAD

**Issue by StrangeTcy** [ageron/handson-ml#28]
In `14_recurrent_neural_networks.ipynb` the cell that's directly after the Deep RNN title didn't work for me.
The error reported was:
```
ValueError                                Traceback (most recent call last)
<ipython-input-41-fcd832f48dd9> in <module>()
     10 basic_cell = tf.contrib.rnn.BasicRNNCell(num_units=n_neurons)
     11 multi_layer_cell = tf.contrib.rnn.MultiRNNCell([basic_cell] * n_layers)
---> 12 outputs, states = tf.nn.dynamic_rnn(multi_layer_cell, X, dtype=tf.float32)
     13 
     14 init = tf.global_variables_initializer()

/usr/local/lib/python3.5/dist-packages/tensorflow/python/ops/rnn.py in dynamic_rnn(cell, inputs, sequence_length, initial_state, dtype, parallel_iterations, swap_memory, time_major, scope)
    551         swap_memory=swap_memory,
    552         sequence_length=sequence_length,
--> 553         dtype=dtype)
    554 
    555     # Outputs of _dynamic_rnn_loop are always shaped [time, batch, depth].

/usr/local/lib/python3.5/dist-packages/tensorflow/python/ops/rnn.py in _dynamic_rnn_loop(cell, inputs, initial_state, parallel_iterations, swap_memory, sequence_length, dtype)
    718       loop_vars=(time, output_ta, state),
    719       parallel_iterations=parallel_iterations,
--> 720       swap_memory=swap_memory)
    721 
    722   # Unpack final output if not using output tuples.

/usr/local/lib/python3.5/dist-packages/tensorflow/python/ops/control_flow_ops.py in while_loop(cond, body, loop_vars, shape_invariants, parallel_iterations, back_prop, swap_memory, name)
   2621     context = WhileContext(parallel_iterations, back_prop, swap_memory, name)
   2622     ops.add_to_collection(ops.GraphKeys.WHILE_CONTEXT, context)
-> 2623     result = context.BuildLoop(cond, body, loop_vars, shape_invariants)
   2624     return result
   2625 

/usr/local/lib/python3.5/dist-packages/tensorflow/python/ops/control_flow_ops.py in BuildLoop(self, pred, body, loop_vars, shape_invariants)
   2454       self.Enter()
   2455       original_body_result, exit_vars = self._BuildLoop(
-> 2456           pred, body, original_loop_vars, loop_vars, shape_invariants)
   2457     finally:
   2458       self.Exit()

/usr/local/lib/python3.5/dist-packages/tensorflow/python/ops/control_flow_ops.py in _BuildLoop(self, pred, body, original_loop_vars, loop_vars, shape_invariants)
   2404         structure=original_loop_vars,
   2405         flat_sequence=vars_for_body_with_tensor_arrays)
-> 2406     body_result = body(*packed_vars_for_body)
   2407     if not nest.is_sequence(body_result):
   2408       body_result = [body_result]

/usr/local/lib/python3.5/dist-packages/tensorflow/python/ops/rnn.py in _time_step(time, output_ta_t, state)
    703           skip_conditionals=True)
    704     else:
--> 705       (output, new_state) = call_cell()
    706 
    707     # Pack state if using state tuples

/usr/local/lib/python3.5/dist-packages/tensorflow/python/ops/rnn.py in <lambda>()
    689 
    690     input_t = nest.pack_sequence_as(structure=inputs, flat_sequence=input_t)
--> 691     call_cell = lambda: cell(input_t, state)
    692 
    693     if sequence_length is not None:

/usr/local/lib/python3.5/dist-packages/tensorflow/contrib/rnn/python/ops/core_rnn_cell_impl.py in __call__(self, inputs, state, scope)
    951                 state, [0, cur_state_pos], [-1, cell.state_size])
    952             cur_state_pos += cell.state_size
--> 953           cur_inp, new_state = cell(cur_inp, cur_state)
    954           new_states.append(new_state)
    955     new_states = (tuple(new_states) if self._state_is_tuple else

/usr/local/lib/python3.5/dist-packages/tensorflow/contrib/rnn/python/ops/core_rnn_cell_impl.py in __call__(self, inputs, state, scope)
    118   def __call__(self, inputs, state, scope=None):
    119     """Most basic RNN: output = new_state = act(W * input + U * state + B)."""
--> 120     with _checked_scope(self, scope or "basic_rnn_cell", reuse=self._reuse):
    121       output = self._activation(
    122           _linear([inputs, state], self._num_units, True))

/usr/lib/python3.5/contextlib.py in __enter__(self)
     57     def __enter__(self):
     58         try:
---> 59             return next(self.gen)
     60         except StopIteration:
     61             raise RuntimeError("generator didn't yield") from None

/usr/local/lib/python3.5/dist-packages/tensorflow/contrib/rnn/python/ops/core_rnn_cell_impl.py in _checked_scope(cell, scope, reuse, **kwargs)
     75             "this error will remain until then.)"
     76             % (cell, cell_scope.name, scope_name, type(cell).__name__,
---> 77                type(cell).__name__))
     78     else:
     79       weights_found = False

ValueError: Attempt to reuse RNNCell <tensorflow.contrib.rnn.python.ops.core_rnn_cell_impl.BasicRNNCell object at 0x7f845c714160> with a different variable scope than its first use.  First use of cell was with scope 'rnn/multi_rnn_cell/cell_0/basic_rnn_cell', this attempt is with scope 'rnn/multi_rnn_cell/cell_1/basic_rnn_cell'.  Please create a new instance of the cell if you would like it to use a different set of weights.  If before you were using: MultiRNNCell([BasicRNNCell(...)] * num_layers), change to: MultiRNNCell([BasicRNNCell(...) for _ in range(num_layers)]).  If before you were using the same cell instance as both the forward and reverse cell of a bidirectional RNN, simply create two instances (one for forward, one for reverse).  In May 2017, we will start transitioning this cell's behavior to use existing stored weights, if any, when it is called with scope=None (which can lead to silent model degradation, so this error will remain until then.)
```
The system used is  Ubuntu 16.04 x64,with both python2 and python3 installed and tensorflow-gpu present for both pythons.

**Comment by jingw222** [comment]
I have similar issue with executing this cell of code, and I got this error:

`ValueError: Trying to share variable rnn/multi_rnn_cell/cell_0/basic_rnn_cell/kernel, but specified shape (200, 100) and found shape (102, 100).`

Why does it expect shape (200, 100) and get shape (102, 100)? Since under the assumption of your previous tutorial explaining things under the hood, It should expect shape (100, 100) instead. Previously we had this:

```
Y0 = tf.tanh(tf.matmul(X0, Wx) + b) # 4*3 3*5 >> 4*5
Y1 = tf.tanh(tf.matmul(Y0, Wy) + tf.matmul(X1, Wx) + b)
```

So there was a `sum` op between the hidden state and a new input. However, the error indicating shape of either (200, 100) or (102, 100) was generally telling me It was a `concat` op instead, and the new input of dimension 2 went into the 'concat' op without multiplying its weight (_(100 + 2 = 102) as opposed to (100 + 100 = 200)_). Don't know if I get myself across.

**Comment by ageron** [comment]
Hi James,
It's a very busy week for me, so I can't investigate right now, but I'll have some time next week. In the meantime, could you please give me more details on your installation? What version of TensorFlow are you using? Can you try installing tensorflow 1.1? There might be some breaking changes in TensorFlow 1.2 regarding RNNs.

**Comment by ageron** [comment]
Ok, I've look into this error, it seems to be due to a breaking change in TensorFlow 1.1 in how stacked RNNs are handled.

With TensorFlow 1.0, this code works fine:

```pycon
>>> import tensorflow as tf
>>> cell = tf.contrib.rnn.BasicRNNCell(num_units=100)
>>> multi_layer_cell = tf.contrib.rnn.MultiRNNCell([cell]*3)
>>> X = tf.placeholder(tf.float32, [None, 20, 10])
>>> outputs, states = tf.nn.dynamic_rnn(multi_layer_cell, X, dtype=tf.float32)
>>> 
```

Then I upgrade to TensorFlow 1.1:

```bash
$ pip install --upgrade tensorflow
[...]
Successfully installed tensorflow-1.1.0
```

And now the same exact code causes the problem:

```pycon
>>> import tensorflow as tf
>>> cell = tf.contrib.rnn.BasicRNNCell(num_units=100)
>>> multi_layer_cell = tf.contrib.rnn.MultiRNNCell([cell]*3)
>>> X = tf.placeholder(tf.float32, [None, 20, 10])
>>> outputs, states = tf.nn.dynamic_rnn(multi_layer_cell, X, dtype=tf.float32)
Traceback (most recent call last):
[...]
  File "/home/ageron/dev/py/envs/ml/lib/python3.5/site-packages/tensorflow/contrib/rnn/python/ops/core_rnn_cell_impl.py", line 77, in _checked_scope
    type(cell).__name__))
ValueError: Attempt to reuse RNNCell <tensorflow.contrib.rnn.python.ops.core_rnn_cell_impl.BasicRNNCell object at 0x7f5887f9cc18> with a different variable scope than its first use.  First use of cell was with scope 'rnn/multi_rnn_cell/cell_0/basic_rnn_cell', this attempt is with scope 'rnn/multi_rnn_cell/cell_1/basic_rnn_cell'.  Please create a new instance of the cell if you would like it to use a different set of weights.  If before you were using: MultiRNNCell([BasicRNNCell(...)] * num_layers), change to: MultiRNNCell([BasicRNNCell(...) for _ in range(num_layers)]).  If before you were using the same cell instance as both the forward and reverse cell of a bidirectional RNN, simply create two instances (one for forward, one for reverse).  In May 2017, we will start transitioning this cell's behavior to use existing stored weights, if any, when it is called with scope=None (which can lead to silent model degradation, so this error will remain until then.)
>>> 
```

So now the proper way to create a `MultiRNNCell` looks like this:

```pycon
>>> import tensorflow as tf
>>> cells = [tf.contrib.rnn.BasicRNNCell(num_units=100) for layer in range(3)]
>>> multi_layer_cell = tf.contrib.rnn.MultiRNNCell(cells)
>>> X = tf.placeholder(tf.float32, [None, 20, 10])
>>> outputs, states = tf.nn.dynamic_rnn(multi_layer_cell, X, dtype=tf.float32)
>>>
```
I'm fixing the notebook for chapter 14, I'll push it by tomorrow.

Thanks for your feedback! :)

**Comment by ageron** [comment]
Ok, I fixed the notebook to support TensorFlow 1.1+.
Thanks again for bringing this issue to my attention. :+1: 

Cheers,
Aurélien

**Comment by StrangeTcy** [comment]
Ok, thanks a bunch!
